sys/include/json.h

```diff
@@ -0,0 +1,859 @@
+ * READ FUNCTIONS
+ *************************************************************************** */
+
+/**
+ * @defgroup    sys_json_read JSON deserialization
+ * @brief       The functions used to deserialize data from a JSON stream.
+ * @{
+ */
+
+/**
+ * @brief   Indicator for a JSON type.
+ *
+ * @see json_read_peek()
+ */
+typedef enum {
+    /**
+     * @brief An object: `{}`
+     * @see json_write_object_open()
+     * @see json_write_object_key()
+     * @see json_write_object_close()
+     * @see json_read_object()
+     * @see json_read_object_next()
+     */
+    JSON_OBJECT,
+
+    /**
+     * @brief An array: `[]`
+     * @see json_write_array_open()
+     * @see json_write_array_next()
+     * @see json_write_array_close()
+     * @see json_read_array()
+     * @see json_read_array_next()
+     */
+    JSON_ARRAY,
+
+    /**
+     * @brief A number: `42`, `47.11`, ...
+     * @see json_write_int()
+     * @see json_write_int64()
+     * @see json_write_float()
+     * @see json_read_number()
+     * @see json_number_type()
+     */
+    JSON_NUMBER,
+
+    /**
+     * @brief A string: `"RIOT"`
+     * @see json_write_string()
+     * @see json_read_string()
+     */
+    JSON_STRING,
+
+    /**
+     * @brief The literal `true`
+     * @see json_write_true()
+     * @see json_read_true()
+     */
+    JSON_TRUE,
+
+    /**
+     * @brief The literal `false`
+     * @see json_write_false()
+     * @see json_read_false()
+     */
+    JSON_FALSE,
+
+    /**
+     * @brief The literal `null`
+     * @see json_write_null()
+     * @see json_read_null()
+     */
+    JSON_NULL,
+} json_type_t;
+
+struct json_read_cookie;
+
+/**
+ * @brief See @ref json_read_cookie.
+ */
+typedef struct json_read_cookie json_read_cookie_t;
+
+/**
+ * @brief   The function used to read a byte from the input stream.
+ *
+ * If your input is a `char` array, then explicitly cast the return value
+ * to `unsigned char` or an 8-bit input will be interpreted as the end of
+ * the stream.
+ *
+ * The result does not differentiate between the end of the stream and the
+ * end of the input.
+ *
+ * @param[in] cookie            The cookie, initialized with json_read_init().
+ *
+ * @returns `0 <= result <= 255`: the read byte.
+ * @returns `result < 0`: the stream ended or there was an error.
+ */
+typedef int (*json_read_t)(json_read_cookie_t *cookie);
+
+/**
+ * @brief The internal state of the input stream.
+ */
+typedef enum {
+    JSON_READ_STATE_SOMEWHERE,   /**< No other state applies. */
+    JSON_READ_STATE_IN_OBJECT,   /**< The last consumed character was `{`. */
+    JSON_READ_STATE_IN_ARRAY,    /**< The last consumed character was `[`. */
+    JSON_READ_STATE_IN_SEQUENCE, /**< The stream is inside a sequence (a key, a string or a number). */
+} json_read_state_t;
+
+/**
+ * @brief   The cookie with all the information needed to deserialize JSON data.
+ *
+ * @see json_read_init()
+ * @see @ref container_of
```

Comment:
Why is this reference here?

Reply:
You'll most likely use this struct as a member of your own struct to keep track of it's state. If `struct json_write_cookie` is the first member, then you could simply cast the pointer, but if not you'll have to use the macro. If you think that's already obvious, then the reference may be removed.

Reply:
I think such behavior should not be encouraged ;-), so it should be removed. See also [this talk](https://media.ccc.de/v/35c3-9670-safe_and_secure_drivers_in_high-level_languages#t=184) ;-).